File: src/spoj/dynamic_programming/ASSIGN.java
package spoj.dynamic_programming;

// http://www.spoj.com/problems/ASSIGN/

import java.util.BitSet;

public class ASSIGN {

	public static void main(String[] args) {

		System.out.println(getAssignmentsNumber(new int[][]{
				{1, 1, 1},
				{1, 1, 1},
				{1, 1, 1}
		}));

		System.out.println(getAssignmentsNumber(new int[][]{
				{1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1},
				{1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0},
				{1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0},
				{1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1},
				{0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1},
				{1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0},
				{0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1},
				{1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1},
				{0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 1},
				{1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1},
				{1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0}
		}));

		System.out.println(getAssignmentsNumber(new int[][]{
				{0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 0},
				{0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1},
				{1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0},
				{0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1},
				{1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1},
				{0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1},
				{1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0},
				{1, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0},
				{0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1},
				{0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1},
				{0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1}
		}));
	}

	static long getAssignmentsNumber(int[][] studentsPreferences) {
		int studentsCount = studentsPreferences.length;
		int tasksCount = studentsPreferences[0].length;

		BitSet[] studentsPreferencesBitSet = new BitSet[studentsCount];
		for (int s = 0; s < studentsCount; s++) {
			studentsPreferencesBitSet[s] = new BitSet(tasksCount);
			for (int t = 0; t < tasksCount; t++) {
				if (studentsPreferences[s][t] == 1) {
					studentsPreferencesBitSet[s].set(t, true);
				}
			}
		}

		BitSet unavailableTasksBitSet = new BitSet(tasksCount);

		return getAssignmentsNumber(studentsPreferencesBitSet, studentsCount - 1, unavailableTasksBitSet);
	}

	// Solution with exponential complexity
	// TODO: memoize[studentIndex][unavailableTasksBitMask]
	// max number of students: 20
	// max number of tasks: 20
	// max value unavailableTasksBitMask is 2^20 = 1'048'576
	static long getAssignmentsNumber(BitSet[] studentsPreferences, int studentIndex, BitSet unavailableTasks) {
		if (studentIndex < 0) {
			return 1;
		}

		long result = 0;
		BitSet currStudentPreferences = studentsPreferences[studentIndex];
		for (int i = 0; i < currStudentPreferences.size(); i++) {
			if (currStudentPreferences.get(i) && !unavailableTasks.get(i)) {
				unavailableTasks.set(i, true);
				result += getAssignmentsNumber(studentsPreferences, studentIndex - 1, unavailableTasks);
				unavailableTasks.set(i, false);
			}
		}

		return result;
	}
}

File: src/spoj/dynamic_programming/ASSIGN.c
// http://www.spoj.com/problems/ASSIGN/

#include <stdio.h>

#define MAX_STUDENTS 20
#define MAX_TASKS 20

long studentsPreferences[MAX_STUDENTS];
// according to description of the problem:
// students count == tasks count
int tasksCount;

// memeory consumption might be decreased: from (N * 2^N) -> to (2 ^ N)
// in case of bottom-up Dynamic Programming solution (instead of top-down solution)
unsigned long long memoized[MAX_STUDENTS][1 << MAX_TASKS];
int is_memoized[MAX_STUDENTS][1 << MAX_TASKS];
int testCaseNum = 0;

inline int isBitSet(long mask, int index) {
    return (mask & (1 << index));
}

inline void toggleBit(long * mask, int index) {
    (*mask) ^= (1 << index);
}

// Complexity is: O((N^2) * (2^N))
unsigned long long calculatePossibleAssignments(int studentIndex, long unavailableTasksMask) {
    if(studentIndex < 0) {
        return 1;
    }
    
    if(is_memoized[studentIndex][unavailableTasksMask] == testCaseNum) {
        return memoized[studentIndex][unavailableTasksMask];
    }
    
    long currentStudentPreferences = studentsPreferences[studentIndex];
    
    unsigned long long result = 0;
    
    int i;
    for(i = 0; i < tasksCount; ++i) {
        if(isBitSet(currentStudentPreferences, i) && !isBitSet(unavailableTasksMask, i)) {
            toggleBit(&unavailableTasksMask, i);
            result += calculatePossibleAssignments(studentIndex - 1, unavailableTasksMask);
            toggleBit(&unavailableTasksMask, i);
        }
    }
    
    memoized[studentIndex][unavailableTasksMask] = result;
    is_memoized[studentIndex][unavailableTasksMask] = testCaseNum;
    
    return result;
}

int main() {
    int testsNum;
    int i;
    int j;
    int isPreferred;
    long unavailableTasksMask;
    
    scanf("%d", &testsNum);
    
    while(testsNum--) {
        scanf("%d", &tasksCount);
        
        for(i = 0; i < tasksCount; ++i) {
            studentsPreferences[i] = 0;
            
            for(j = 0; j < tasksCount; ++j) {
                scanf("%d", &isPreferred);
                
                if(isPreferred) {
                    toggleBit(&studentsPreferences[i], j);
                }
            }
        }
        
        unavailableTasksMask = 0;
        testCaseNum++;
        
        printf("%llu\n", calculatePossibleAssignments(tasksCount - 1, unavailableTasksMask));
    }
    return 0;
}